app/models/katello/concerns/pulp_database_unit.rb

```diff
@@ -175,15 +181,41 @@ def copy_repository_associations(source_repo, dest_repo)
                     select #{dest_repo.id} as repository_id, pulp_id, #{columns.join(',')} from #{self.table_name}
                     where repository_id = #{source_repo.id} and pulp_id not in (select pulp_id
                     from #{self.table_name} where repository_id = #{dest_repo.id})"
+          ActiveRecord::Base.connection.execute(insert_query)
 
         end
         ActiveRecord::Base.connection.execute(delete_query)
-        ActiveRecord::Base.connection.execute(insert_query)
       end
 
       def with_pulp_id(unit_pulp_ids)
         where('pulp_id in (?)', unit_pulp_ids)
       end
+
+      def db_columns_sync
+        [unit_id_field.to_sym, backend_identifier_field, :repository_id, :created_at, :updated_at].compact
+      end
+
+      def db_columns_copy
+        [self.unit_id_field.to_sym, backend_identifier_field, :repository_id].compact
+      end
+
+      def db_values_copy(source_repo, dest_repo)
+        db_values = []
+        source_units = self.repository_association_class.where(:repository_id => source_repo.id)
+        dest_units = self.repository_association_class.where(:repository_id => dest_repo.id)
+        source_unit_ids = source_units.pluck(unit_id_field.to_sym)
+        dest_unit_ids = dest_units.pluck(unit_id_field.to_sym)
+        new_units = source_units.select { |su| (source_unit_ids - dest_unit_ids).include? su[unit_id_field] }
```

Comment:
Would it be better to move `(source_unit_ids - dest_unit_ids)` into a variable out of the select loop?

```
missing_unit_ids = source_unit_ids - dest_unit_ids
new_unit_db_values = source_units.map do |su|
  next unless missing_unit_ids.include?(su[unit_id_field])
  [unit[unit_id_field], unit[backend_identifier_field], dest_repo.id].compact
end
new_unit_db_values.compact
```

Reply:
Hmm. Is there a way to find the missing units with a single active record query rather than running two queries and taking the difference?